File: data_structures/5_kotlin/j_hashTables.kt
class HashTable<Key, Value>(private val capacity: Int) {

    private val buckets: ArrayList<ArrayList<Pair<Key, Value>>> = ArrayList(capacity)

    init {
        require(capacity > 0) { "Capacity should be greater than 0" }

        // Initialize the buckets
        repeat(capacity) { buckets.add(ArrayList()) }
    }

    private fun bucketIndex(key: Key): Int {
        val hash = key.hashCode()
        return hash.rem(capacity)
    }

    fun getValue(key: Key): Value? {
        val index = bucketIndex(key)
        for (pair in buckets[index]) {
            if (pair.first == key) {
                return pair.second
            }
        }
        return null
    }

    fun setValue(key: Key, value: Value) {
        val index = bucketIndex(key)

        // Check if the key already exists, and update the value
        for (i in buckets[index].indices) {
            if (buckets[index][i].first == key) {
                buckets[index][i] = Pair(key, value)
                return
            }
        }

        // Key doesn't exist, add a new entry
        buckets[index].add(Pair(key, value))
    }

    fun removeValue(key: Key) {
        val index = bucketIndex(key)
        buckets[index].removeIf { it.first == key }
    }

    fun removeAll() {
        for (bucket in buckets) {
            bucket.clear()
        }
    }

    fun count(): Int {
        return buckets.sumBy { it.size }
    }

    fun contains(key: Key): Boolean {
        val index = bucketIndex(key)
        return buckets[index].any { it.first == key }
    }

    fun allKeys(): List<Key> {
        val keys = ArrayList<Key>()
        for (bucket in buckets) {
            for (pair in bucket) {
                keys.add(pair.first)
            }
        }
        return keys
    }

    fun allValues(): List<Value> {
        val values = ArrayList<Value>()
        for (bucket in buckets) {
            for (pair in bucket) {
                values.add(pair.second)
            }
        }
        return values
    }

    fun merge(otherTable: HashTable<Key, Value>) {
        for (bucket in otherTable.buckets) {
            for ((key, value) in bucket) {
                setValue(key, value)
            }
        }
    }

    fun resize(newCapacity: Int) {
        require(newCapacity > 0) { "New capacity should be greater than 0" }

        val newBuckets: ArrayList<ArrayList<Pair<Key, Value>>> = ArrayList(newCapacity)

        repeat(newCapacity) { newBuckets.add(ArrayList()) }

        for (bucket in buckets) {
            for ((key, value) in bucket) {
                val newIndex = bucketIndex(key)
                newBuckets[newIndex].add(Pair(key, value))
            }
        }

        buckets.clear()
        buckets.addAll(newBuckets)
    }

    // ... Add more hash table operations as needed ...
}

fun main() {
    val table = HashTable<Int, String>(10)

    table.setValue(1, "Value1")
    table.setValue(2, "Value2")
    table.setValue(3, "Value3")

    println("Value for key 2: ${table.getValue(2)}")

    table.removeValue(2)
    println("Count after removal: ${table.count()}")

    println("Keys in the table: ${table.allKeys()}")
}

File: data_structures/1_cpp/j_hashTables.cpp
#include <iostream>
#include <vector>
#include <functional>

template <typename Key, typename Value>
class HashTable {
private:
    using Element = std::pair<Key, Value>;
    using Bucket = std::vector<Element>;

    std::vector<Bucket> buckets;

    size_t totalBuckets() const {
        return buckets.size();
    }

public:
    // Initialization
    HashTable(size_t capacity) {
        assert(capacity > 0 && "Capacity should be greater than 0");
        buckets = std::vector<Bucket>(capacity);
    }

    // Access
    Value* getValue(const Key& key) {
        size_t index = bucketIndex(key);
        for (const Element& element : buckets[index]) {
            if (element.first == key) {
                return &buckets[index][element];
            }
        }
        return nullptr;
    }

    void setValue(const Value& value, const Key& key) {
        size_t index = bucketIndex(key);

        // Check if the key already exists, and update the value
        for (auto& element : buckets[index]) {
            if (element.first == key) {
                element.second = value;
                return;
            }
        }

        // Key doesn't exist, add a new entry
        buckets[index].emplace_back(key, value);
    }

    // Removal
    void removeValue(const Key& key) {
        size_t index = bucketIndex(key);
        buckets[index].erase(std::remove_if(buckets[index].begin(), buckets[index].end(),
            [key](const Element& element) { return element.first == key; }), buckets[index].end());
    }

    void removeAll() {
        for (auto& bucket : buckets) {
            bucket.clear();
        }
    }

    // Helpers
    size_t bucketIndex(const Key& key) const {
        return std::hash<Key>{}(key) % totalBuckets();
    }

    // Count
    size_t count() const {
        size_t totalCount = 0;
        for (const auto& bucket : buckets) {
            totalCount += bucket.size();
        }
        return totalCount;
    }

    // Check Existence
    bool contains(const Key& key) const {
        size_t index = bucketIndex(key);
        for (const auto& element : buckets[index]) {
            if (element.first == key) {
                return true;
            }
        }
        return false;
    }

    // Keys and Values
    std::vector<Key> allKeys() const {
        std::vector<Key> keys;
        for (const auto& bucket : buckets) {
            for (const auto& element : bucket) {
                keys.push_back(element.first);
            }
        }
        return keys;
    }

    std::vector<Value> allValues() const {
        std::vector<Value> values;
        for (const auto& bucket : buckets) {
            for (const auto& element : bucket) {
                values.push_back(element.second);
            }
        }
        return values;
    }

    // Merging
    void merge(const HashTable<Key, Value>& otherTable) {
        for (const auto& bucket : otherTable.buckets) {
            for (const auto& element : bucket) {
                setValue(element.second, element.first);
            }
        }
    }

    // Resizing
    void resize(size_t newCapacity) {
        // Create a new hash table with the desired capacity
        HashTable<Key, Value> newHashTable(newCapacity);

        // Merge the elements from the current hash table
        newHashTable.merge(*this);

        // Assign the new hash table
        *this = newHashTable;
    }
};

// ... Add more hash table operations as needed ...

int main() {
    HashTable<int, std::string> table(10);

    table.setValue("Value1", 1);
    table.setValue("Value2", 2);
    table.setValue("Value3", 3);

    std::cout << "Value for key 2: " << *(table.getValue(2)) << "\n";

    table.removeValue(2);
    std::cout << "Count after removal: " << table.count() << "\n";

    std::vector<int> keys = table.allKeys();
    std::cout << "Keys in the table:";
    for (const auto& key : keys) {
        std::cout << " " << key;
    }
    std::cout << "\n";

    return 0;
}
